### PR TITLE
Add round trip tests of Property Assignment methods

### LINKED ISSUE #1481: Feature:  Create Round-trip property tests
Generate new test cases to convert each API type to the associated storage type and back again, verifying that the round trip is lossless. Necessary to ensure that all of this generated conversion code actually works as intended. The use of property generators should help us find any odd edge cases.

Given the assumption that we have only only API version of each service for `codegen-alpha-0`, this will give complete coverage. There are other tests we'll need once we have more than one API version.

 Uses the generator framework that already exists for testing that individual types can round trip to JSON and back again.

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Adds tests to test that the `AssignPropertiesTo*()` and `AssignPropertiesFrom*()` methods are symmetric and copy all the properties across between instances. These tests are property based, and each is run as a batch.

Sample output:
```
PS> go test -run WhenPropertiesConverted
+ Round trip from BatchAccount to BatchAccount via AssignPropertiesToBatchAccount & AssignPropertiesFromBatchAccount returns original: OK, passed 100 tests.
+ Round trip from BatchAccount_Status to BatchAccount_Status via AssignPropertiesToBatchAccountStatus & AssignPropertiesFromBatchAccountStatus returns original: OK, passed 100 tests.
+ Round trip from BatchAccounts_Spec to BatchAccounts_Spec via AssignPropertiesToBatchAccountsSpec & AssignPropertiesFromBatchAccountsSpec returns original: OK, passed 100 tests.
+ Round trip from AutoStorageBaseProperties to AutoStorageBaseProperties via AssignPropertiesToAutoStorageBaseProperties & AssignPropertiesFromAutoStorageBaseProperties returns original: OK, passed 100 tests.
+ Round trip from AutoStorageProperties_Status to AutoStorageProperties_Status via AssignPropertiesToAutoStoragePropertiesStatus & AssignPropertiesFromAutoStoragePropertiesStatus returns original: OK, passed 100 tests.
+ Round trip from BatchAccountIdentity to BatchAccountIdentity via AssignPropertiesToBatchAccountIdentity & AssignPropertiesFromBatchAccountIdentity returns original: OK, passed 100 tests.
+ Round trip from BatchAccountIdentity_Status to BatchAccountIdentity_Status via AssignPropertiesToBatchAccountIdentityStatus & AssignPropertiesFromBatchAccountIdentityStatus returns original: OK, passed 100 tests.
+ Round trip from EncryptionProperties to EncryptionProperties via AssignPropertiesToEncryptionProperties & AssignPropertiesFromEncryptionProperties returns original: OK, passed 100 tests.
```

Closes #1481

**Prerequisites**

- [x] Merge PR #1727 
- [ ] Rebase on `master`

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: hack/generator/pkg/testcases/property_assignment_test_case.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+	"github.com/dave/dst"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astbuilder"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/conversions"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/functions"
+)
+
+// PropertyAssignmentTestCase represents a test that verifies we can convert from one object in our converstion graph
```
Comment (babbageclunk) [inline hack/generator/pkg/testcases/property_assignment_test_case.go:20]:
```suggestion
// PropertyAssignmentTestCase represents a test that verifies we can convert from one object in our conversion graph
```

File: hack/generator/pkg/codegen/pipeline/testdata/TestInjectPropertyAssignmentTests/microsoft.person-v20211231_test.golden [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+)
+
+func Test_Person_WhenPropertiesConverted_RoundTripsWithoutLoss(t *testing.T) {
+	parameters := gopter.DefaultTestParameters()
+	parameters.MaxSize = 10
+	properties := gopter.NewProperties(parameters)
+	properties.Property(
+		"Round trip from Person to Person via AssignPropertiesToPerson & AssignPropertiesFromPerson returns original",
+		prop.ForAll(RunPropertyAssignmentTestForPerson, PersonGenerator()))
+	properties.TestingRun(t, gopter.NewFormatedReporter(false, 240, os.Stdout))
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/testdata/TestInjectPropertyAssignmentTests/microsoft.person-v20211231_test.golden:28]: Every time I see "formated" my eye twitches. I'm guessing you feel the same way.

**Review by babbageclunk — APPROVED** [review]
This looks good, but I'ma bit  worried about how many tests we're going to be generating/running if this is what comes from a couple of versions of a pretty small type. (I guess this is the same for the json tests?)

File: hack/generator/pkg/codegen/pipeline/testdata/TestInjectPropertyAssignmentTests/microsoft.person-v20211231_test.golden (hunk 2 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/codegen/pipeline/testdata/TestInjectPropertyAssignmentTests/microsoft.person-v20211231_test.golden:28]: Yup. 

gif

File: hack/generator/pkg/testcases/property_assignment_test_case.go (hunk 1 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/testcases/property_assignment_test_case.go:20]: Thanks.

**Review by theunrepentantgeek — COMMENTED** [review]
> This looks good, but I'm a bit worried about how many tests we're going to be generating/running if this is what comes from a couple of versions of a pretty small type. (I guess this is the same for the json tests?)

Most of the infrastructure is shared with the JSON tests, so the amount of new stuff is actually pretty low. @matthchr has some ideas on reducing the amount of code, which I'll look into.
